compiler/passes/normalize.cpp

```diff
@@ -4530,8 +4530,17 @@ static void updateInitMethod(FnSymbol* fn) {
     preNormalizeInitMethod(fn);
 
   } else if (thisType == dtUnknown) {
-    INT_FATAL(fn, "'this' argument has unknown type");
-
+    if (ArgSymbol* thisArg = toArgSymbol(fn->_this)) {
+      if (thisArg->typeExpr->body.length == 1) {
+        Expr* expr = thisArg->typeExpr->body.only();
+        if (UnresolvedSymExpr* se = toUnresolvedSymExpr(expr)) {
+          USR_FATAL(fn, "initializer defined on unrecognized type: '%s'",
+                    se->unresolved);
+          return;
+        }
+      }
+    }
+    USR_FATAL(fn, "initializer defined on unrecognized type");
```

Comment:
Could this have been a `USR_FATAL_CONT`? A similar error below does that.

Reply:
That's a good idea, and I've done it, but I don't know whether it's a net win or loss.  E.g., is getting two errors for these lines better than one?  Is the old error sufficiently good that we should just make this step return rather than printing a special error?

Reply:
To be clear, the one above returns already, so we shouldn't get double errors (right ?).

Reply:
Ah, now I see what you mean. And now that you made that comment I pulled your branch to test something :)

```chapel
var x: R;

proc R.init() {

}

writeln(x);
```

If you remove this error altogether, the compiler will only complain about the variable declaration, which seems a bit unfortunate, but probably because that error is not `_CONT`...

For this particular case, `USR_WARN` is appealing to me. But in your case, it'll be a warning *and* an error. Which is marginally less redundant than what you pasted above.

We'll need some error caching mechanism as the ideal solution here. (Somewhat related: https://github.com/Cray/chapel-private/issues/1095). For now, my preference in order is:

1. USR_WARN here
2. Do nothing here
3. Any USR_FATAL here (I can't tell which is better at this point)

Reply:
> but probably because that error is not _CONT...

No, that error is a _CONT.  I think it's because of the "first use this function" aspect.  I.e., if you refer to `R` 100 times within a code block, we print 1 error rather than 100.  For example, note that if you reverse the order of those two declarations, we still error on the first occurrence even though it's now the `R.init`.

Reply:
Hmm, I see. My order is more like 1=2>3, then.

Reply:
OK, I think my preference is 2 if, by 2, you mean "remove the previous INT_FATAL and just continue compilation until the next error catches it."  I think the benefits are that we fix the bug, lean more on existing error infrastructure (such that when/as it improves, this case improves with it), and avoid putting in this fiddly code to handle this specific case somewhat arbitrarily (i.e., why shouldn't `proc R.foo()` or `R + 1` get similar consideration and special error messages?).

Let me know if you're OK with that and I'll proceed with it.

Reply:
Yep, that sounds good.